File: backend/src/scripts/init-scan.ts
#!/usr/bin/env node
/**
 * Git 仓库初始化扫描脚本
 * 
 * 用法:
 *   pnpm init-scan              # 扫描最近3个月
 *   pnpm init-scan --months 6   # 扫描最近6个月
 *   pnpm init-scan --months 12  # 扫描最近12个月
 */

// 加载环境变量
import 'dotenv/config';

import simpleGit, { SimpleGit } from 'simple-git';
import { RepositoryService } from '../services/RepositoryService.js';
import { ConfigService } from '../services/ConfigService.js';
import { GitScanService } from '../services/GitScanService.js';
import { CommitService } from '../services/CommitService.js';
import { logger } from '../config/logger.js';

/**
 * 解析命令行参数
 */
function parseArgs(): { months: number } {
  const args = process.argv.slice(2);
  let months = 3; // 默认3个月

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--months' && args[i + 1]) {
      const value = parseInt(args[i + 1], 10);
      if (!isNaN(value) && value > 0 && value <= 24) {
        months = value;
      } else {
        logger.warn(`无效的月份参数: ${args[i + 1]}，使用默认值 3 个月`);
      }
      break;
    }
  }

  return { months };
}

/**
 * 更新 Git 仓库代码（执行 git pull）
 */
async function pullRepository(repoPath: string, repoName: string): Promise<boolean> {
  try {
    const git: SimpleGit = simpleGit(repoPath);
    
    // 检查是否为有效的 Git 仓库
    const isRepo = await git.checkIsRepo();
    if (!isRepo) {
      logger.warn(`  [Git Pull] ${repoName} 不是有效的 Git 仓库，跳过更新`);
      return false;
    }

    // 检查是否有远程仓库配置
    const remotes = await git.getRemotes(true);
    if (remotes.length === 0) {
      logger.info(`  [Git Pull] ${repoName} 没有配置远程仓库，跳过更新`);
      return false;
    }

    logger.info(`  [Git Pull] 正在更新 ${repoName}...`);
    
    // 执行 git pull
    const pullResult = await git.pull();
    
    if (pullResult.summary.changes > 0 || pullResult.summary.insertions > 0 || pullResult.summary.deletions > 0) {
      logger.info(`  [Git Pull] ✓ ${repoName} 更新成功: ${pullResult.summary.changes} 个文件变更, +${pullResult.summary.insertions}/-${pullResult.summary.deletions}`);
    } else {
      logger.info(`  [Git Pull] ✓ ${repoName} 已是最新版本`);
    }
    
    return true;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    logger.warn(`  [Git Pull] ⚠ ${repoName} 更新失败: ${errorMessage}，将继续扫描`);
    // 不抛出错误，允许继续扫描
    return false;
  }
}

/**
 * 将时间范围分割成多个周
 */
function splitIntoWeeks(startDate: Date, endDate: Date): Array<{ start: Date; end: Date }> {
  const weeks: Array<{ start: Date; end: Date }> = [];
  let currentStart = new Date(startDate);
  
  while (currentStart < endDate) {
    const currentEnd = new Date(currentStart);
    currentEnd.setDate(currentEnd.getDate() + 7); // 一周
    
    // 确保不超过结束日期
    const weekEnd = currentEnd > endDate ? endDate : currentEnd;
    
    weeks.push({
      start: new Date(currentStart),
      end: weekEnd
    });
    
    currentStart = new Date(weekEnd);
    // 避免无限循环，如果日期相同则前进一天
    if (currentStart.getTime() === weekEnd.getTime()) {
      currentStart.setDate(currentStart.getDate() + 1);
    }
  }
  
  return weeks;
}

/**
 * 分阶段扫描单个仓库（支持断点续传）
 */
async function scanRepositoryInPhases(
  repoConfig: { id: string; name: string; path: string },
  targetStartDate: Date, // 目标起始日期
  endDate: Date,
  authorEmails: string[],
  repositoryService: RepositoryService,
  commitService: CommitService
): Promise<{ scanned: number; inserted: number; skipped: number; completed: boolean }> {
  // 获取当前扫描进度
  const currentScanToDate = await repositoryService.getInitialScanToDate(repoConfig.id);

  // 确定实际开始日期：如果有进度，从未完成的日期继续；否则从目标日期开始
  const actualStartDate = currentScanToDate
    ? new Date(currentScanToDate)
    : targetStartDate;

  if (actualStartDate >= endDate) {
    logger.info(`  [已完成] 已扫描到 ${new Date(currentScanToDate!).toISOString().split('T')[0]}，无需继续扫描`);
    return { scanned: 0, inserted: 0, skipped: 0, completed: true };
  }

  // 只扫描未完成的部分
  const weeks = splitIntoWeeks(actualStartDate, endDate);
  let totalScanned = 0;
  let totalInserted = 0;
  let totalSkipped = 0;
  let lastScannedDate = actualStartDate;

  logger.info(`  [分阶段扫描] 从 ${actualStartDate.toISOString().split('T')[0]} 继续，共 ${weeks.length} 个阶段`);

  for (let i = 0; i < weeks.length; i++) {
    const week = weeks[i];
    const weekNumber = i + 1;

    try {
      logger.info(`  [阶段 ${weekNumber}/${weeks.length}] ${week.start.toISOString().split('T')[0]} ~ ${week.end.toISOString().split('T')[0]}`);

      // 执行扫描（使用日期范围限制，支持多作者）
      const scanner = new GitScanService(repoConfig.path);
      // 将结束日期设置为下一周的开始，确保包含本周的所有提交
      const weekEndPlusOne = new Date(week.end);
      weekEndPlusOne.setDate(weekEndPlusOne.getDate() + 1);
      const commits = await scanner.incrementalScan(week.start, authorEmails, weekEndPlusOne);

      if (commits.length > 0) {
        // 统计分支信息
        const branches = new Set(commits.map(c => c.branch).filter((b): b is string => !!b));
        const branchInfo = branches.size > 0 
          ? `，涉及 ${branches.size} 个分支: ${Array.from(branches).slice(0, 5).join(', ')}${branches.size > 5 ? '...' : ''}`
          : '';
        
        // 保存提交记录（带去重，支持多分支）
        const insertResult = await commitService.batchInsertCommits(repoConfig.id, commits);
        
        logger.info(`    ✓ 发现: ${commits.length} 条${branchInfo}, 新增: ${insertResult.inserted} 条, 跳过: ${insertResult.skipped} 条`);

        totalScanned += commits.length;
        totalInserted += insertResult.inserted;
        totalSkipped += insertResult.skipped;
      } else {
        logger.info(`    - 无提交记录`);
      }

      // 更新扫描进度（记录到本周结束日期）
      lastScannedDate = week.end;
      await repositoryService.updateInitialScanToDate(repoConfig.id, week.end.getTime());
      logger.info(`    [进度更新] 已扫描到 ${week.end.toISOString().split('T')[0]}`);

      // 每个阶段之间稍作延迟，避免过载
      if (i < weeks.length - 1) {
        await new Promise(resolve => setTimeout(resolve, 100));
      }
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error(`  [阶段 ${weekNumber} 失败] ${errorMessage}`);
      // 即使失败，也更新进度到当前阶段的开始日期，下次从这个阶段重新开始
      await repositoryService.updateInitialScanToDate(repoConfig.id, week.start.getTime());
      logger.info(`    [进度保存] 失败后保存进度到 ${week.start.toISOString().split('T')[0]}，下次从此继续`);
      // 继续下一个阶段，不中断整个扫描
    }
  }

  // 检查是否已完成（扫描到当前时间）
  const completed = lastScannedDate.getTime() >= endDate.getTime();

  return {
    scanned: totalScanned,
    inserted: totalInserted,
    skipped: totalSkipped,
    completed
  };
}

/**
 * 初始化扫描
 */
async function initScan() {
  const { months } = parseArgs();
  const configService = new ConfigService();
  const repositoryService = new RepositoryService();
  const commitService = new CommitService();
  
  logger.info('='.repeat(60));
  logger.info(`[初始化扫描] 开始扫描最近 ${months} 个月的提交记录（支持断点续传）...`);
  logger.info('='.repeat(60));

  const targetDate = new Date();
  targetDate.setMonth(targetDate.getMonth() - months);
  const now = new Date();
  const targetDateTimestamp = targetDate.getTime();

  let totalScanned = 0;
  let totalInserted = 0;
  let totalSkipped = 0;
  let successCount = 0;
  let failCount = 0;
  let completedCount = 0;

  // 获取所有启用的仓库配置
  const repos = await configService.getEnabledRepositories();
  logger.info(`[初始化] 找到 ${repos.length} 个启用的仓库\n`);

  for (const repo of repos) {

    // 检查该仓库是否已完成初始扫描
    const isCompleted = await repositoryService.isInitialScanCompleted(repo.id, targetDateTimestamp);
    if (isCompleted) {
      logger.info(`[跳过] ${repo.name} (已完成初始扫描到目标日期)`);
      completedCount++;
      continue;
    }

    try {
      logger.info(`\n[扫描仓库] ${repo.name}`);
      logger.info(`  路径: ${repo.path}`);
      
      // 先更新仓库代码到最新
      await pullRepository(repo.path, repo.name);
      
      // 获取当前扫描进度
      const currentScanToDate = await repositoryService.getInitialScanToDate(repo.id);
      if (currentScanToDate) {
        logger.info(`  继续扫描: 从 ${new Date(currentScanToDate).toISOString().split('T')[0]} 继续`);
      } else {
        logger.info(`  首次扫描: 从 ${targetDate.toISOString().split('T')[0]} 开始`);
      }
      logger.info(`  目标日期: ${targetDate.toISOString().split('T')[0]}`);

      // 获取该仓库的作者邮箱列表
      const authorEmails = await configService.getAuthorEmailsByRepoId(repo.id);
      if (authorEmails.length === 0) {
        logger.warn(`  [跳过] 仓库 ${repo.name} 没有配置作者，跳过`);
        continue;
      }

      // 分阶段扫描（支持断点续传）
      const result = await scanRepositoryInPhases(
        repo,
        targetDate,
        now,
        authorEmails,
        repositoryService,
        commitService
      );

      totalScanned += result.scanned;
      totalInserted += result.inserted;
      totalSkipped += result.skipped;

      // 更新仓库信息
      const dbResult = await commitService.getCommits({
        startDate: 0,
        endDate: Date.now(),
        repositoryIds: [repo.id],
        page: 1,
        pageSize: 1
      });

      await repositoryService.updateRepositoryScanInfo(
        repo.id,
        Date.now(),
        dbResult.total
      );

      if (result.completed) {
        logger.info(`  [完成] 已扫描到目标日期，仓库总计: ${dbResult.total} 条提交记录`);
        completedCount++;
      } else {
        logger.info(`  [部分完成] 已保存进度，仓库总计: ${dbResult.total} 条提交记录，下次继续扫描`);
      }
      
      successCount++;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error(`[扫描失败] ${repo.name}: ${errorMessage}`);
      failCount++;
      // 继续扫描下一个仓库，不中断整个流程
    }
  }

  logger.info('\n' + '='.repeat(60));
  logger.info('[扫描完成] 统计信息:');
  logger.info(`  已完成仓库: ${completedCount} 个`);
  logger.info(`  进行中仓库: ${successCount - completedCount} 个`);
  logger.info(`  失败仓库: ${failCount} 个`);
  logger.info(`  扫描提交总数: ${totalScanned} 条`);
  logger.info(`  新增入库: ${totalInserted} 条`);
  logger.info(`  跳过重复: ${totalSkipped} 条`);
  logger.info('='.repeat(60) + '\n');

  process.exit(0);
}

// 运行初始化扫描
initScan().catch((error) => {
  const errorMessage = error instanceof Error ? error.message : String(error);
  const errorStack = error instanceof Error ? error.stack : undefined;
  logger.error({
    msg: 'Initial scan failed',
    error: errorMessage,
    stack: errorStack
  });
  process.exit(1);
});

File: python-backend/app/scripts/init_scan.py
#!/usr/bin/env python
"""Git 仓库初始化扫描脚本

用法:
  python -m app.scripts.init_scan              # 扫描最近3个月
  python -m app.scripts.init_scan --months 6   # 扫描最近6个月
  python -m app.scripts.init_scan --months 12  # 扫描最近12个月

支持断点续传：每个仓库按周分阶段扫描，进度保存在数据库中，
下次运行会从上次中断的位置继续。
"""

from __future__ import annotations

import argparse
import asyncio
import sys
import time
from datetime import datetime, timedelta, timezone
from pathlib import Path

from loguru import logger

# ---------------------------------------------------------------------------
# CLI argument parsing
# ---------------------------------------------------------------------------


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Git 仓库初始化扫描")
    parser.add_argument(
        "--months",
        type=int,
        default=3,
        help="扫描最近 N 个月的提交记录 (默认 3, 最大 24)",
    )
    return parser.parse_args()


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _pull_repository(repo_path: str, repo_name: str) -> bool:
    """Try ``git pull`` on the given repo. Returns ``True`` on success."""
    import git as gitpython

    try:
        repo = gitpython.Repo(repo_path)

        if not repo.remotes:
            logger.info(f"  [Git Pull] {repo_name} 没有配置远程仓库，跳过更新")
            return False

        logger.info(f"  [Git Pull] 正在更新 {repo_name}...")
        remote = repo.remotes.origin
        pull_info = remote.pull()

        changes = sum(1 for info in pull_info if info.flags != 0) if pull_info else 0
        if changes > 0:
            logger.info(f"  [Git Pull] ✓ {repo_name} 更新成功")
        else:
            logger.info(f"  [Git Pull] ✓ {repo_name} 已是最新版本")
        return True
    except Exception as exc:
        logger.warning(f"  [Git Pull] ⚠ {repo_name} 更新失败: {exc}，将继续扫描")
        return False


def _split_into_weeks(
    start_date: datetime, end_date: datetime
) -> list[tuple[datetime, datetime]]:
    """Split a date range into weekly chunks."""
    weeks: list[tuple[datetime, datetime]] = []
    current_start = start_date

    while current_start < end_date:
        current_end = current_start + timedelta(days=7)
        week_end = min(current_end, end_date)
        weeks.append((current_start, week_end))

        current_start = week_end
        # avoid infinite loop
        if current_start == week_end:
            current_start += timedelta(days=1)

    return weeks


# ---------------------------------------------------------------------------
# Phase-based scan (checkpoint / resume)
# ---------------------------------------------------------------------------


async def _scan_repo_in_phases(
    repo: dict,
    target_start: datetime,
    end_date: datetime,
    author_emails: list[str],
) -> dict:
    """Scan a single repository in weekly phases with checkpoint support.

    Returns ``{"scanned": int, "inserted": int, "skipped": int, "completed": bool}``.
    """
    from app.core.database import async_session_factory
    from app.services.commit_service import CommitService
    from app.services.git_scan_service import GitScanService
    from app.services.repository_service import RepositoryService

    repo_id = repo["id"]
    repo_name = repo["name"]

    async with async_session_factory() as db:
        # Get current scan progress
        scan_to_date_ms = await RepositoryService.get_initial_scan_to_date(db, repo_id)

    actual_start = (
        datetime.fromtimestamp(scan_to_date_ms / 1000, tz=timezone.utc)
        if scan_to_date_ms
        else target_start
    )

    if actual_start >= end_date:
        logger.info(
            f"  [已完成] 已扫描到 {actual_start.strftime('%Y-%m-%d')}，无需继续扫描"
        )
        return {"scanned": 0, "inserted": 0, "skipped": 0, "completed": True}

    weeks = _split_into_weeks(actual_start, end_date)
    total_scanned = 0
    total_inserted = 0
    total_skipped = 0
    last_scanned_date = actual_start

    logger.info(
        f"  [分阶段扫描] 从 {actual_start.strftime('%Y-%m-%d')} 继续，"
        f"共 {len(weeks)} 个阶段"
    )

    scanner = GitScanService(repo["path"])

    for i, (week_start, week_end) in enumerate(weeks, 1):
        try:
            logger.info(
                f"  [阶段 {i}/{len(weeks)}] "
                f"{week_start.strftime('%Y-%m-%d')} ~ {week_end.strftime('%Y-%m-%d')}"
            )

            # Extend end by 1 day to include all commits in the week
            week_end_plus_one = week_end + timedelta(days=1)
            commits = await scanner.incremental_scan(
                week_start, author_emails, to_date=week_end_plus_one
            )

            if commits:
                branches = {
                    c.branch for c in commits if c.branch
                }
                branch_info = ""
                if branches:
                    sample = list(branches)[:5]
                    suffix = "..." if len(branches) > 5 else ""
                    branch_info = f"，涉及 {len(branches)} 个分支: {', '.join(sample)}{suffix}"

                commit_dicts = [c.to_dict() for c in commits]

                async with async_session_factory() as db:
                    insert_result = await CommitService.batch_insert_commits(
                        db, repo_id, commit_dicts
                    )
                    await db.commit()

                logger.info(
                    f"    ✓ 发现: {len(commits)} 条{branch_info}, "
                    f"新增: {insert_result['inserted']} 条, "
                    f"跳过: {insert_result['skipped']} 条"
                )

                total_scanned += len(commits)
                total_inserted += insert_result["inserted"]
                total_skipped += insert_result["skipped"]
            else:
                logger.info("    - 无提交记录")

            # Update checkpoint
            last_scanned_date = week_end
            week_end_ms = int(week_end.timestamp() * 1000)
            async with async_session_factory() as db:
                await RepositoryService.update_initial_scan_to_date(
                    db, repo_id, week_end_ms
                )
                await db.commit()

            logger.info(
                f"    [进度更新] 已扫描到 {week_end.strftime('%Y-%m-%d')}"
            )

            # Small delay between phases
            if i < len(weeks):
                await asyncio.sleep(0.1)

        except Exception as exc:
            logger.error(f"  [阶段 {i} 失败] {exc}")
            # Save progress to the start of this phase so we retry from here
            week_start_ms = int(week_start.timestamp() * 1000)
            async with async_session_factory() as db:
                await RepositoryService.update_initial_scan_to_date(
                    db, repo_id, week_start_ms
                )
                await db.commit()
            logger.info(
                f"    [进度保存] 失败后保存进度到 {week_start.strftime('%Y-%m-%d')}，下次从此继续"
            )

    completed = last_scanned_date >= end_date
    return {
        "scanned": total_scanned,
        "inserted": total_inserted,
        "skipped": total_skipped,
        "completed": completed,
    }


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------


async def init_scan(months: int) -> None:
    from app.core.database import async_session_factory
    from app.services.commit_service import CommitService
    from app.services.config_service import ConfigService
    from app.services.repository_service import RepositoryService

    logger.info("=" * 60)
    logger.info(
        f"[初始化扫描] 开始扫描最近 {months} 个月的提交记录（支持断点续传）..."
    )
    logger.info("=" * 60)

    now = datetime.now(tz=timezone.utc)
    target_start = (now - timedelta(days=months * 30)).replace(
        hour=0, minute=0, second=0, microsecond=0
    )
    target_start_ms = int(target_start.timestamp() * 1000)

    total_scanned = 0
    total_inserted = 0
    total_skipped = 0
    success_count = 0
    fail_count = 0
    completed_count = 0

    # Get all enabled repos
    async with async_session_factory() as db:
        repos = await ConfigService.get_enabled_repos(db)
    logger.info(f"[初始化] 找到 {len(repos)} 个启用的仓库\n")

    for repo in repos:
        repo_id = repo["id"]
        repo_name = repo["name"]

        # Check if initial scan is already completed
        async with async_session_factory() as db:
            is_done = await RepositoryService.is_initial_scan_completed(
                db, repo_id, target_start_ms
            )

        if is_done:
            logger.info(f"[跳过] {repo_name} (已完成初始扫描到目标日期)")
            completed_count += 1
            continue

        try:
            logger.info(f"\n[扫描仓库] {repo_name}")
            logger.info(f"  路径: {repo['path']}")

            # Pull latest code
            _pull_repository(repo["path"], repo_name)

            # Show scan progress
            async with async_session_factory() as db:
                scan_to_date_ms = await RepositoryService.get_initial_scan_to_date(
                    db, repo_id
                )

            if scan_to_date_ms:
                progress_date = datetime.fromtimestamp(
                    scan_to_date_ms / 1000, tz=timezone.utc
                )
                logger.info(
                    f"  继续扫描: 从 {progress_date.strftime('%Y-%m-%d')} 继续"
                )
            else:
                logger.info(
                    f"  首次扫描: 从 {target_start.strftime('%Y-%m-%d')} 开始"
                )
            logger.info(f"  目标日期: {now.strftime('%Y-%m-%d')}")

            # Get author emails
            async with async_session_factory() as db:
                author_emails = await ConfigService.get_author_emails_by_repo_id(
                    db, repo_id
                )

            if not author_emails:
                logger.warning(f"  [跳过] 仓库 {repo_name} 没有配置作者，跳过")
                continue

            # Phase-based scan
            result = await _scan_repo_in_phases(
                repo, target_start, now, author_emails
            )

            total_scanned += result["scanned"]
            total_inserted += result["inserted"]
            total_skipped += result["skipped"]

            # Update repo scan info
            async with async_session_factory() as db:
                count_result = await CommitService.get_commits(
                    db,
                    start_date=0,
                    end_date=int(time.time() * 1000),
                    repository_ids=[repo_id],
                    page=1,
                    page_size=1,
                )
                await RepositoryService.update_repo_scan_info(
                    db, repo_id, int(time.time() * 1000), count_result["total"]
                )
                await db.commit()

            if result["completed"]:
                logger.info(
                    f"  [完成] 已扫描到目标日期，"
                    f"仓库总计: {count_result['total']} 条提交记录"
                )
                completed_count += 1
            else:
                logger.info(
                    f"  [部分完成] 已保存进度，"
                    f"仓库总计: {count_result['total']} 条提交记录，下次继续扫描"
                )

            success_count += 1
        except Exception as exc:
            logger.error(f"[扫描失败] {repo_name}: {exc}")
            fail_count += 1

    logger.info("\n" + "=" * 60)
    logger.info("[扫描完成] 统计信息:")
    logger.info(f"  已完成仓库: {completed_count} 个")
    logger.info(f"  进行中仓库: {success_count - completed_count} 个")
    logger.info(f"  失败仓库: {fail_count} 个")
    logger.info(f"  扫描提交总数: {total_scanned} 条")
    logger.info(f"  新增入库: {total_inserted} 条")
    logger.info(f"  跳过重复: {total_skipped} 条")
    logger.info("=" * 60 + "\n")


def main() -> None:
    args = parse_args()
    months = args.months

    if months < 1 or months > 24:
        logger.warning(f"无效的月份参数: {months}，使用默认值 3 个月")
        months = 3

    try:
        asyncio.run(init_scan(months))
    except KeyboardInterrupt:
        logger.info("[中断] 用户中断扫描")
        sys.exit(0)
    except Exception as exc:
        logger.error(f"初始化扫描失败: {exc}")
        sys.exit(1)


if __name__ == "__main__":
    main()
